enhancement(sinks): Adaptive Request Concurrency

bruceg opened the pull request:
Here is my implementation of the automatic concurrency management, which is in use in all "service2" based sinks (which appears to be everything except prometheus and aws_cloudwatch_logs).

I will continue to work on the tests started in `auto_concurrency/service.rs`, but otherwise I believe this is feature complete.

Closes #2529

Commit:
Import base concurrency limiter from tower

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop unused Load impl on AutoConcurrencyLimit

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce Controller wrapper for the semaphore

The implementation is complicated by the fact that we can only decrease
concurrency by acquiring permits and forgetting them. If there aren't
enough free permits, this requires waiting for them when polling for
readiness.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Adjust pub markers and drop more unused methods

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Link the semaphore controller and start time into the response future

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Calculate (and throw away) RTT in response future

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add function to controller to calculate average RTT

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add increase and decrease methods to the controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Initial implementation of RTT-based concurrency adjustment

Has issues that need correcting:

* has recursive locking via `fn contract` and `fn expand`
* adjusts on every measurement, not once per RTT
* adjusts for any response, does not differentiate

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix recursive locking when making adjustments, and rename "dropping"

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Limit adjustments to once per RTT

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce new EWMA abstraction

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Aggregate responses for each RTT interval

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Simplify retry logic with `matches!`

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Wrap the shrinkable semaphore in its own module

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Use new stub AutoConcurrencyLimit layer in service2-based sinks

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add RetryLogic to auto concurrency layer (unused)

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Handle back pressure through applying RetryLogic (BROKEN)

I think the problem is that the Timeout layer needs to be inside the
AutoConcurrencyLimit layer, but the Error type in the Timeout layer is a
Box<Error> which differs from the RetryLogic Error type.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Hoverbear committed:
Fixup ;)

Signed-off-by: Ana Hobden <operator@hoverbear.org>

Commit:
Start work on tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix: Make sure there is an initial set of measurements

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix: Only reset the current measurements after using them

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Start of auto request limiting test framework

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Silence a clippy warning in tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move `get_controller` and `capture_metrics` into src/metrics.rs

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move creation of Metric from a Measurement into `src/event/metric.rs`

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Introduce an `assert_between` macro for range matches

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rename `current` to `current_limit` in controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add and test internal metrics for the controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move autoreq test parameters into their own struct

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Generate the test statistics all in one go, and add mode

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rename assert_between to assert_within

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix handling of error responses in the controller

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add tests for dropped and deferred requests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add get_ref methods to access internals of our service layers

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Make test_util into a module

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Move autoreq tests into the src tree beside auto_concurrency

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Update Cargo for new get_ref methods in tokio03

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Drop unused use std::error::Error

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add get_mut functions

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Enhance the test_util stats

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Replace use of metrics with controller internal stats

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add a version of assert_within! that adds extra output

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add impl Display for the WeightedSum types

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Increase the boundary threshold to 5%

Small jitter in timings was preventing concurrency increases from
happening.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
More test enhancements

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix and test handling of senders that never reach the limit

Don't activate the concurrency increases if the limiter doesn't see
requests beyond the current concurrency limit. This prevents the limiter
from scaling up concurrency beyond what it has evidence for.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add test for a medium-speed sender

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix some test parameters

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add test for RTT jitter

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Convert only usage of tokio01 in tests to new tokio

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Some more test tweaks

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Use tokio's Instant to determine "now" for testing

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Start of some trivial low-level service tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add Rust 2018 edition tag to .rustfmt.toml

Some editors use `rustfmt` directly instead of `cargo fmt`, which
requires explicitly indicating which Rust edition the project is using.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
More service level tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,234 @@
+        let rtt = now.saturating_duration_since(start);
+        emit!(AutoConcurrencyObservedRtt { rtt });
+        let rtt = rtt.as_secs_f64();
+        #[cfg(test)]
```
Please add a newline above this one

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,234 @@
+        }
+
+        #[cfg(test)]
+        stats.in_flight.add(inner.in_flight, now);
```
Please add a newline below this one

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,234 @@
+    fn adjust_to_back_pressure(&self, start: Instant, is_back_pressure: bool) {
+        let now = instant_now();
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+        #[cfg(test)]
```
Please add a newline above this one

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,234 @@
+}
+
+#[derive(Clone, Copy, Debug, Default)]
+struct EWMA {
```
```suggestion
/// Exponentially Weighted Moving Average
struct EWMA {
```

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,234 @@
+}
+
+impl Mean {
+    fn update(&mut self, point: f64) -> f64 {
```
```suggestion
    /// Update and return the current average
    fn update(&mut self, point: f64) -> f64 {
```

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,234 @@
+    fn update(&mut self, point: f64) -> f64 {
+        self.sum += point;
+        self.count += 1;
+        // Return current average
```
```suggestion
```

Hoverbear commented on src/sinks/util/auto_concurrency/semaphore.rs:
```diff
@@ -0,0 +1,93 @@
+    future: Pin<Box<dyn Future<Output = OwnedSemaphorePermit> + Send + 'static>>,
+}
+
+impl Future for MaybeForgetFuture {
```
I think this is a really slick method @bruceg

bruceg replied:
It took me a few iterations to realize that kind of solution.

Commit:
Overrule a clippy lint, no good way to do await in a closure

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Improve internal controller documentation

Signed-off-by: Bruce Guenter <bruce@timber.io>

fanatid reviewed (commented):
Impressive work.

Commit:
Merge remote-tracking branch 'origin/master' into auto-concurrency

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Another quick clippy fix

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/auto_concurrency/semaphore.rs:
```diff
@@ -0,0 +1,93 @@
+            let future = self.master.semaphore.clone().acquire_owned();
+            replace(&mut self.future, Box::pin(future));
+        }
+        drop(to_forget);
```
`to_forget` is a lock guard and, while the following call to `poll` *should* be fast, we don't want to hold the lock for any longer than necessary to prevent blocking elsewhere, so I did an explicit drop as soon as it is not needed.

fanatid replied:
Thanks for explanation. It was strange for me to see `drop` before `poll` here, while lock guard is held during `poll` in the loop.

bruceg commented on src/sinks/util/auto_concurrency/semaphore.rs:
```diff
@@ -0,0 +1,93 @@
+/// number of permits before yielding a valid one.
+struct MaybeForgetFuture {
+    master: Arc<ShrinkableSemaphore>,
+    future: Pin<Box<dyn Future<Output = OwnedSemaphorePermit> + Send + 'static>>,
```
I was probably "programming by error message" when I put that one together. I'll look through and fix that.

Commit:
Simplify semaphore struct with BoxFuture

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/auto_concurrency/semaphore.rs:
```diff
@@ -0,0 +1,93 @@
+            .to_forget
+            .lock()
+            .expect("Shrinkable semaphore mutex is poisoned");
+        for _ in 0..count {
+            match self.semaphore.try_acquire() {
+                Ok(permit) => permit.forget(),
+                Err(_) => *to_forget += 1,
+            }
+        }
```
My thinking was that another thread could release a permit while the loop is running, and then subsequent iterations might not all have the error.

bruceg commented on src/sinks/util/auto_concurrency/mod.rs:
```diff
@@ -0,0 +1,15 @@
+pub(crate) use service::AutoConcurrencyLimit;
+
+pub(self) fn instant_now() -> std::time::Instant {
+    tokio::time::Instant::now().into()
```
This is for testing done in `auto_concurrency/service.rs`. It allows me to pause and manually advance the test clock without delaying the program.

bruceg commented on src/sinks/util/sink.rs:
```diff
@@ -157,6 +157,16 @@ where
     }
 }
 
+impl<S, B, R, E> BatchSink<S, B, R, E> {
+    pub fn get_ref(&self) -> &S {
+        &self.service.service
+    }
+
+    pub fn get_mut(&mut self) -> &mut S {
+        &mut self.service.service
+    }
+}
+
```
Most of the tower layers support that kind of access to the inner service, which is the only way my tests can share (`Arc::clone`) the inner statistics. I only really need the `fn get_ref`, and they could be guarded with `#[cfg(test)]`, but there is precedent in tower.

fanatid replied:
I meant that `get_ref`/`get_mut` should return `ServiceSink<S, Request>`, not `S` (i.e. 1 level)? From the other side I see now that `new` method expects `S`, not `ServiceSink<S, Request>`...

bruceg commented on src/sinks/util/auto_concurrency/service.rs:
```diff
@@ -0,0 +1,319 @@
+        .await;
+    }
+
+    #[allow(clippy::needless_range_loop)]
```
I don't think we have a policy, so it would be a useful discussion to have elsewhere. My thinking is to leave as many clippy lints on as possible, to allow all of the code to continue to benefit from them, and turn them off in specific places where we have hand verified they are incorrect. FTR I would like to avoid this one too, but couldn't figure out a way to do it without a range loop (still discussing in Slack though).

Commit:
Add comment about clippy lint allow in service test

Signed-off-by: Bruce Guenter <bruce@timber.io>

Hoverbear commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,240 @@
+            emit!(AutoConcurrencyAveragedRtt {
+                rtt: Duration::from_secs_f64(rtt)
+            });
+            #[cfg(test)]
```
This should be nudged down to help the reader notice the flag. :)

Hoverbear reviewed (commented):
Do you think it would be valuable to do a bench run on this? Do a criterion baseline on the main branch and compare this? https://bheisler.github.io/criterion.rs/book/user_guide/command_line_options.html#baselines

binarylogic commented:
Just noting a couple of blocking requirements:

- [ ] We need someone on the team to take this for a spin and try out scenarios where this would prevent a downstream outage.
- [ ] I want to verify that this does not affect benchmarks. Basically ensuring that the warm-up period is no worse than what we currently have. That might be as simple as ensuring default/starting values remain the same.

Commit:
More blank lines around #[cfg(test)] sections

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into auto-concurrency

Signed-off-by: Bruce Guenter <bruce@timber.io>

lukesteensen commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+            else if inner.current_limit < self.max
+                && inner.reached_limit
+                && !inner.had_back_pressure
+                && rtt <= avg
```
This `rtt <= avg` strikes me as a little bit conservative and I'm curious what the effect would be if we gave it some wiggle room (e.g. within a 2-5% of the current average or below).

Very much out of scope for this PR, but it'd be interesting to have a little simulator to test things out like that.

bruceg replied:
Actually, this is a little big aggressive, as this is the logic for _increasing_ the concurrency limit. This was the purpose of `THRESHOLD_RATIO`, but I only used it the other case. I went back and forth on this, so I'm open to tweaking the threshold logic.

I definitely agree it would be useful to have a simulation setup to test this. One of my thoughts was to work in both `EWMA_ALPHA` and `THRESHOLD_RATIO` as undocumented global parameters so we could test using an actual vector build, but I didn't work that into this (already large) PR.

lukesteensen commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+        response: &Result<L::Response, crate::Error>,
+    ) {
+        let is_back_pressure = match response {
+            Ok(_) => false,
```
We may want to add a method to `RetryLogic` similar to `should_retry_response` that looks for things like HTTP 429 where we're explicitly being told to slow down. Not necessary for this PR, but could be a good follow-up.

bruceg replied:
Definitely, yes. At one point I had created an `IsBackPressure` trait (or something like that) that would be used in place of `RetryLogic` on response and error types, but discarded that approach.

lukesteensen commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+                } else if err.downcast_ref::<Elapsed>().is_some() {
+                    true
+                } else {
+                    panic!("Unhandled error response! {:?}", err)
```
Do we have anything that will prevent this panic at compile time? I don't want to get into a situation where we're panicking because we forgot to implement a certain error type.

bruceg replied:
No, unfortunatly there is nothing we _can_ do at compile time to prevent this panic. This downcast is only required because the error type of the `Timeout` layer is `Box<Error>` instead of an enum. The only errors that `Timeout` produces is either the input error type or `Elapsed`, so it should not be possible to trigger the panic, but I can't guarantee that.

I did consider rewriting the `Timeout` layer as well so that it's `type Error` was an actual enum instead of a `Box`, but figured that was out of scope for this as well. I'm not happy with this solution.

binarylogic replied:
> I did consider rewriting the Timeout layer as well so that it's type Error was an actual enum instead of a Box, but figured that was out of scope for this as well. I'm not happy with this solution.

Should we open this as a tech-debt issue?

bruceg replied:
I would, but it's a tech-debt issue in the `tower` crate, which would also introduce a breaking API change. How does that factor in?

binarylogic replied:
Ah. So it's unlikely to be accepted? If that's the case it sounds like it's not worth it, but I am not a good judge of the benefit it would unlock.

bruceg replied:
I would guess it to be unlikely to be accepted, yes, but you never know. The benefits are avoiding a needless `Box` and enforcing type-correctness on users of the timeout layer error type, but I don't know how the tower crew would perceive those benefits.

binarylogic replied:
It sounds like it's worth opening an issue on that repo first to see if the change would be accepted. If you agree, would you mind doing that and referencing it here?

bruceg replied:
Will do.

bruceg replied:
Opened https://github.com/tower-rs/tower/issues/465

lukesteensen replied:
I wonder if we could do something tricky where we bundle the inner service, retry logic, this layer, and timeout layer into one struct (or even constructor fn) where we can bound the inner `Service::Error` to the `RetryLogic::Error`. That wouldn't get rid of the box in the middle, but it'd help prevent us from constructing a mismatched stack.

bruceg replied:
I'm sorry, this isn't tricky enough? :astonished:

lukesteensen replied:
Just brainstorming :smile: 

If you're comfortable with the current risk level, that's good enough for me!

bruceg replied:
It turns out tower used to do exactly what we want, but it had issues too, particularly when layers were nested. See https://github.com/tower-rs/tower/issues/131

MOZGIII replied:
Based on the discussion above, it seems like this can be changed to `unreachable!("unhandled error response! {:?}", err)` to better capture the intent of the branch.

bruceg replied:
Agreed.

lukesteensen commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+impl Mean {
+    /// Update and return the current average
+    fn update(&mut self, point: f64) -> f64 {
+        self.sum += point;
+        self.count += 1;
```
Do you think we need to worry about overflow for either of these?

bruceg replied:
I doubt it. `sum` is a `f64`, and it would be pretty astonishing for that to overflow in human time scales. Similarly, `count` is `usize`, which will take a rather impossibly large number of packets to overflow.

lukesteensen commented on src/test_util/mod.rs:
```diff
@@ -30,6 +32,68 @@ macro_rules! assert_downcast_matches {
 }
 
+#[macro_export]
+macro_rules! assert_within {
```
We actually already have a similar crate pulled in for this: https://github.com/timberio/vector/blob/122778d90ffb7644e6811543696c845a1fb0417d/src/transforms/sampler.rs#L109

bruceg replied:
While I could use that macro instead of this, the semantics are different (within is `Low <= Value <= High`, relative is `abs(Ideal - Actual) <= Epsilon`) I can rewrite it though if you think it's worth it.

lukesteensen replied:
Nope, not a big deal. Just wanted to point it out in case you hadn't seen.

lukesteensen reviewed (commented):
This is looking really cool! I'm excited to see what it can do.

Commit:
Use explicit Arc::clone on Arc objects

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+        if inner.in_flight >= inner.current_limit {
+            inner.reached_limit = true;
+        }
+        self.semaphore.clone().acquire()
```
Ref also #3186. I will fix.

MOZGIII reviewed (commented):
This PR is very interesting! I've only looked through the code so far, and planning to look deeper into the core algorithm in the next few days.

Is there any literature I can read up on this particular algorithm? I wonder how it compares to Token Bucket/Leaky Bucket - I'm familiar with those, but haven't seen an application of moving averages to concurrency control before. I like the approach, but want to know more about it's properties. I'll try to figure out some of the things using the queue theory.

Also, it would be great if the code had more doc-comments, explaining the intents behind the API signatures. It will make the code easier to navigate.


UPD: found a lot of background info at the pointers at the [Prior Art section of the RFC](https://github.com/timberio/vector/blob/master/rfcs/2020-04-06-1858-automatically-adjust-request-limits.md#prior-art).

bruceg commented on src/sinks/util/auto_concurrency/future.rs:
```diff
@@ -0,0 +1,58 @@
+where
+    F: Future<Output = Result<L::Response, E>>,
+    L: RetryLogic,
+    E: Into<crate::Error>,
```
I agree it is at best a bit odd. The short story is that it breaks without it. Longer story is that `AutoConcurrencyLayer` wraps `Timeout`, and `type Timeout::Error` is equivalent to `crate::Error`, which drives some of our type decisions. This was the only way we could get it to work with the trait system, and we (@lukesteensen, @Hoverbear, and I) spent many hours banging at it. If you can figure a better solution I would love to do it better.

Commit:
Fix extraneous module path in moved function in metrics

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Ignore result of metrics::init() with `let _`

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+use tokio::sync::OwnedSemaphorePermit;
+use tower03::timeout::error::Elapsed;
+
+const EWMA_ALPHA: f64 = 0.5;
```
I'll add some notes, but basically there is no rationale other than something reasonable to start testing.

Commit:
Comment on the default EWMA_ALPHA and THRESHOLD_RATIO values

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,243 @@
+    }
+
+    /// Update and return the current average
+    fn update(&mut self, point: f64) -> f64 {
```
Can do.

Commit:
Drop unnecessary `max` in Mean::update

The count is incremented in the line before, so it will always be at
least one, which was what the max was put in place to ensure.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add simple stats tests

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented on src/sinks/util/auto_concurrency/future.rs:
```diff
@@ -0,0 +1,58 @@
+};
+use tokio::sync::OwnedSemaphorePermit;
+
+/// Future for the `AutoConcurrencyLimit` service.
```
Those are good questions. First, rate limiting is wrapped around this one, so it won't see requests that are delayed due to such a limit. Second, holding (not dropping) it for too long can cause a concurrency slot to be used up longer than necessary. As long as it is polled promptly, the computations will be reasonably accurate, and AFAIUI such polling is taken care of by the future being spawned immediately in the sink.

Commit:
Add a bit of docs to the auto future

Signed-off-by: Bruce Guenter <bruce@timber.io>

binarylogic reviewed (commented):
@bruceg very nice work on this. We discussed this on Slack, but I wanted to put it here for posterity. We'd like to change this to accept an `"auto"` value for the `in_flight_limit` option:

```toml
in_flight_limit = "auto"
```

For a few reasons:

1. It's obvious to users that we are automatically adjusting this limit.
2. It allows users to opt-out of this feature by setting it to a static limit if they are having problems with it.
3. A core reason we built this feature was to remove the decision entirely from the user.

To avoid extreme cases, we'd like to place a hard limit of `200` when `in_flight_limit` is set to `"auto"`.

Commit:
Change interpretation of in_flight_limit to a fixed limit

If `request.in_flight_limit` is set, it specifies a fixed constant
limit, effectively bypassing the automatic concurrency mechanisms. If
unset, the limit is allowed to float up to a maximum of MAX_CONCURRENCY,
currently set to 200.

Note that this explicitly does not apply to the aws_s3 and
gcp_cloud_storage sinks, which use the old fixed concurrency limiter,
nor aws_cloudwatch_logs, which uses the old sink service.

Signed-off-by: Bruce Guenter <bruce@timber.io>

MOZGIII commented on src/sinks/util/auto_concurrency/controller.rs:
```diff
@@ -0,0 +1,282 @@
+        let mut inner = self.inner.lock().expect("Controller mutex is poisoned");
+
+        #[cfg(test)]
+        let mut stats = self.stats.lock().expect("Stats mutex is poisoned");
```
Might be useful to switch to `try_lock` here if possible - to avoid the effects of additional locking affecting test results.

bruceg replied:
So, if we use `lock`, then the stats may not reflect how the sink will operate in practice because there is additional lock contention. If we use `try_lock`, then the stats may not reflect how the sink actually did operate because it sometimes didn't take the lock and so couldn't report stats, resulting in incomplete data. Either way, stats can't always reflect actual operation, which I agree is unfortunate. Using `try_lock` also requires additional conditionals when trying to report stats. I think I'll stick with `lock` unless there is another consideration.

Now, I can reorganize the code a bit to make the time the mutex is held even shorter, which will reduce the effect. Do you have any thoughts on whether it is better to take the mutex once and hold it for a bit longer in `adjust_to_back_pressure`, or to take it once for each place it is needed and drop it immediately?

MOZGIII replied:
Yeah, I see the issue with the `try_lock` access to stats.

I had some thoughts on using lock-free ring buffers (channels?) to collect the stats for test purposes, and then to aggregate them separately. But it's very early - I'm still reading through the code and wrapping my head around what's going on.

A bit preliminary, but to me, implementing channel based mocks to tightly control the execution in a tight lock-step looks promising here. This would allow us to assert the states in all the right places, and eliminate the need for manual locking and associated race conditions.

The idea is we use a channel to send notifications that we reached a certain state to the test controlling logic along the way, and we then wait on another rx channel to continue. The test logic waits for a particular notification, does the assertion, and sends a message that lets the tested code continue the execution.

Again, this is very preliminary - I'll have to take a deeper look if this has any actual benefits over what we have already - but I'm very curious to hear what are your thoughts on this.

bruceg replied:
I had considered using channels for reporting test metrics, but couldn't see a path that was as straightforward as using locking. It is worth considering, though.

I see there being roughly 3 levels of testing: low level (does it take the steps that the code says it should), behavior (do the steps give it the expected high-level behavior), and performance (does it interoperate optimally with live servers). It sounds like you are describing some fairly low-level tests, covering some of the steps tested in `auto_concurrency/service.rs` using tokio's timers to manually advance steps. It might be worth covering other paths though.

MOZGIII replied:
Yes, you're right, the approach I'm thinking about is just briefly above, and very similar to, tracing the code along it's execution in lock-step, and asserting that exactly what's expected happens.

I'm thinking of tokio timers as just one of the ways to advance.
Consider having `state` at the test end rather than as part of the struct, and only passing it into the `Controller` code via a channel. That channel will pass ownership and access to `state`, and it'll have to naturally be passed between the threads to mix the updates and assertions.

MOZGIII commented:
I ran tests locally and got some failures.

The command I used was `TEST_LOG=trace cargo test --lib -- sinks::util::auto_concurrency --nocapture 2>&1 | tee log.txt`.

Output: [log.txt](https://github.com/timberio/vector/files/5025972/log.txt)

Commit:
Mark the unhandled error type as `unreachable!`

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
Hmm, I can definitely confirm the stats issues you are having, as I am seeing them now too. When I added the tests, though, I ran them many times to ensure the boundaries were reasonable. While it would be good to track down what changed, and to adjust the boundaries to ensure tests now pass, this makes me wonder if these tests are reliable enough to keep around.

Commit:
Reorganize controller test stats access to reduce lock hold time

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Rename a couple variables to make the algorithm clearer

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Make the automatic limit opt-in by default

This restores the previous defaults for the fixed concurrency limits,
and enables the automatic limiter if `request.in_flight_limit = "auto"`.

This fixes the self tests that were accidentally broken by commit
8605d64b41ca3174ef6540cb7bff9bb1ef355042 changing the behavior of
specifying an in_flight_limit number in the settings.

Signed-off-by: Bruce Guenter <bruce@timber.io>

bruceg commented:
The issue is that commit 8605d64b41ca3174ef6540cb7bff9bb1ef355042 switched the behavior of specifying a number for `in_flight_limit`, effectively turning the tests on to a fixed concurrency again instead of the variable concurrency. I discovered this while working on making the variable concurrency opt-in, and I have a fix.

MOZGIII reviewed (approved):
Looks good overall! We might want to revisit tests though.

Commit:
Really fix concurrency when in_flight_limit is not auto

The previous cheap fix to make in_flight_limit an opt-in parameter did
not properly keep the concurrency limit fixed. In the case of increasing
RTT, it will still decrease the concurrency at some point, and so return
to the variable concurrency behavior. This change completely turns off
the varying mechanism when in_flight_limit is not auto.

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Add a test for fixed concurrency and expand some more result ranges

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into auto-concurrency

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix auto_concurrency tests for new tokio

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Fix clippy lint issues with cloning Arc refs

Signed-off-by: Bruce Guenter <bruce@timber.io>

Commit:
Merge remote-tracking branch 'origin/master' into auto-concurrency

Signed-off-by: Bruce Guenter <bruce@timber.io>

lukesteensen reviewed (approved):
Looks good! Excited to get this in and collect some real-world feedback.

Commit:
Only run the tests on unix systems, due to timing variability problems

Ideally these should run with exact timing, but I haven't worked that in
yet.

Signed-off-by: Bruce Guenter <bruce@timber.io>
